lrpar/src/lib/ctbuilder.rs

```diff
@@ -990,7 +943,50 @@ where
             YaccKind::Eco => unreachable!(),
         };
 
-        let parse_fn = quote! {
+        // `parse()` may or may not have an argument for `%parseparam`.
+        let parse_fn_parse_param = match self.yacckind.unwrap() {
+            YaccKind::Original(YaccOriginalActionKind::UserAction) | YaccKind::Grmtools => {
+                match grm.parse_param() {
+                    Some((name, tyname)) => {
+                        let name = str::parse::<TokenStream>(name)?;
+                        let tyname = str::parse::<TokenStream>(tyname)?;
+                        quote! {#name: #tyname}
+                    }
+                    None => quote!(),
+                }
+            }
+            YaccKind::Original(YaccOriginalActionKind::GenericParseTree) => quote!(),
+            YaccKind::Original(YaccOriginalActionKind::NoAction) => quote!(),
+            YaccKind::Eco => unreachable!(),
+        };
+        let parse_fn_return_ty = match self.yacckind.unwrap() {
+            YaccKind::Original(YaccOriginalActionKind::UserAction) | YaccKind::Grmtools => {
+                let actiont = grm
+                    .actiontype(self.user_start_ridx(grm))
+                    .as_ref()
+                    .map(|at| str::parse::<TokenStream>(at))
+                    .transpose()?;
+                quote! {
+                    (::std::option::Option<#actiont>, ::std::vec::Vec<::lrpar::LexParseError<#storaget, #lexertypest>>)
+                }
+            }
+            YaccKind::Original(YaccOriginalActionKind::GenericParseTree) => quote! {
+                (::std::option::Option<::lrpar::Node<<#lexertypest as ::lrpar::LexerTypes>::LexemeT, #storaget>>,
+                    ::std::vec::Vec<::lrpar::LexParseError<#storaget, #lexertypest>>)
+            },
+            YaccKind::Original(YaccOriginalActionKind::NoAction) => quote! {
+                ::std::vec::Vec<::lrpar::LexParseError<#storaget, #lexertypest>>
+            },
+            YaccKind::Eco => unreachable!(),
+        };
+        let grm_const_name = format_ident!("{}", GRM_CONST_NAME);
+        let stable_const_name = format_ident!("{}", STABLE_CONST_NAME);
+        let grm_data = encode_to_vec(grm, bincode::config::legacy())?;
+        let stable_data = encode_to_vec(stable, bincode::config::legacy())?;
+        Ok(quote! {
+            const #grm_const_name: &[u8] = &[#(#grm_data,)*];
+            const #stable_const_name: &[u8] = &[#(#stable_data,)*];
```

Comment:
At this point, all the uses of these `__GRM_DATA` AND `__STABLE_DATA` consts,
which the `GRM_CONST_NAME` and `STABLE_CONST_NAME` variables refer to are within a few lines of one another.

Perhaps we should just replace these `#grm_const_name` variables with `__GRM_DATA` and remove the
indirection.

Reply:
> Perhaps we should just replace these #grm_const_name variables with __GRM_DATA directly and remove the indirection?

Works for me!

Reply:
Should be fixed in [e2243a6](https://github.com/softdevteam/grmtools/pull/510/commits/e2243a6b0d1dd38c024af11126d91b0307489461)

Overall I'm pretty happy with how this has turned out at this point, and I would be happy with either
unmarking as draft for review, *or* pulling out just the `gen_parse_function` changes.

It don't believe the `output_file` changes when moving the rest of the `gen_` function's to returning `Result<TokenStream,..>`,  would be anything more than one line changes, so I'd be happy to proceed in
whichever fashion you prefer.